libcontainer/nsenter/nsexec.c

```diff
@@ -632,6 +638,76 @@ void join_namespaces(char *nslist)
 	free(namespaces);
 }
 
+/*
+ * enable_ima_ns - sets up IMA Namespace for the process
+ *
+ * In order to set it up correctly calling process
+ * should have new userns first, because if it is in the first userns then
+ * new imans won't be created (imans has no CLONE_* flags). The function does next things:
+ *
+ * 1. It mounts temporary securityfs, in order to not bother hosts's
+ * filesystem, we've chosen container's rootfs as a location for this,
+ * (to be more accurate we've chosen /mnt inside the rootfs). Make
+ * sure the calling process already created new userns.
+ *
+ * 2. It writes '1' to rootfs/mnt/integrity/ima/active. This is
+ * the key stage of enabling imans. After that imans will be actually created.
+ *
+ * 3. Unmount temporary securityfs.
+ *
+ */
+#define IMA_ACTIVE_PATH	"integrity/ima/active"
+/* IMA  securityfs relative mount point */
+#define IMA_SECURITYFS_REL_MNT	"/mnt/"
+
+static void enable_ima_ns(void)
+{
+	int err, mntflags, ima_active_path_len;
+	char data, ima_active[PATH_MAX + 1];
+
+	ima_active_path_len = strlen(IMA_ACTIVE_PATH);
+	/*
+	 * Now current working directory is our rootfs,
+	 * so in order to avoid some conflicts with Host filesystem,
+	 * we should mount securityfs inside the rootfs of the currently
+	 * starting container. The mount point for this securityfs is
+	 * CWD + "/mnt"
+	 */
+	if(!getcwd(ima_active, PATH_MAX))
+		bail("failed to get path of the current rootfs");
+
+	if ((strnlen(ima_active, PATH_MAX) + strlen(IMA_SECURITYFS_REL_MNT)) > PATH_MAX)
+		bail("securityfs temporary path is too long");
+
+	strncat(ima_active, IMA_SECURITYFS_REL_MNT, PATH_MAX);
+	mntflags = MS_NOATIME | MS_NODEV | MS_NOEXEC | MS_NOSUID | MS_RELATIME;
+	err = mount("securityfs", ima_active, "securityfs", mntflags , NULL);
```

Comment:
Does this code assumes that there is a `mnt` directory under container rootfs? This depends on the distro and might not always be the case.

Reply:
You're right this depends on a distro, but most of them follows FHS (Filesystem Hierarchy Standard) which includes `/mnt` directory. Anyway, now I see such a solution:
1. Check whether rootfs contains `/mnt` directory (may be any other directory)
2. If not - create it
3. Do mount securityfs and enable IMA namespace
4. Unmount securityfs
5. Remove `/mnt` directory
6. otherwise use `/mnt` as it is already presented

Another way I see is to create a temporary directory like `.secfs` within current rootfs, and after enabling namespace remove it. But I guess this may affect perfomance as we invoke disk I/O.

Thanks!